services/containerregistry/accesstoken.go

```diff
@@ -0,0 +1,119 @@
+package containerregistry
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ExchangeAccessToken exchanges an Azure Container Registry refresh token for an Azure Container Registry access token
+func (c *ContainerRegistryClient) ExchangeAccessToken(ctx context.Context, refreshToken string, scopes AccessTokenScopes) (string, *AccessTokenClaims, error) {
```

Comment:
It looks like this can be used as an authorizer in itself, would it be reasonable to implement `auth.Authorizer` with this struct? Along those lines, I think it's good to work with oauth2.Token wherever possible as this helps improve interoperability - can we return an `*oauth2.Token` here instead of a string?

Reply:
That would be the best way! But important that it doesn't get confused with the Azure AD authorizers, since it won't do much good over there. Will take a look and see if I can come up with something good! 👍

Reply:
Regarding authorizer: I've been looking at `auth.Authorizer` as a way to obtain Azure AD tokens, not tokens for Azure Container Registry. In this case, it could be confusing to implement the interface since it's also used by itself to obtain AAD tokens.

Regarding oauth2.Token: I'm not sure Azure Container Registry is implementing the flow in a way that's supported by oauth2.Token. But to be honest, I haven't used it that much myself yet. I would be able to populate it with content, but not actually sure if it would serve any greater purpose since we wouldn't be able to use it for refreshing the token (as far as I understand).

Any feedback on the above two points is appriciated. I'm not sure about them and it may be worth you looking at it a bit deeper to see if it's applicable or not - I may be missing something here.

Reply:
In my mind, `oauth2.Token` is just a familiar container type for HTTP authorization tokens, with some helper methods, which is accepted by many packages. 

The `auth.Authorizer` interface is likewise intended to be somewhat generic, guaranteeing that types will be able to supply a primary token for the Authorization header, and any supplementary tokens where applicable for the `X-Ms-Authorization-Auxiliary` header. I think it's fine to accept `Authorizer`s for use in other clients like ACR - how the token is obtained and what it's valid for is largely orthogonal to that :)

If a client wishes to carry out some validation on tokens provided by an Authorizer to determine suitability, parsing the claims should yield sufficient data for that - e.g. comparing the `aud` claim.